### DISCUSSION TITLE
Access child-collection through Interface on parent object (als interface)

### DISCUSSION THREAD

**Discussion by JurjendeGroot** [MarimerLLC/csla#2777] [Questions]
I'm in a project right now were we use the dataportal (webApi) and we use the regular csla objects there. The 'layer' where we actually use our CSLA objects is a .NET standard project so we rely heavily on interfaces. We now have an interface IPerson (used on the readonly and editable object) and IPersonEditable (for the editable object)) 

However, when the BusinessObject is a parent, the child-collections are not available, so when for instance Person would have a property DocumentCollection and AddressCollection these are (now) not available on IPersonEditable, I could add that like so

IEnumerable<IDocumentEditable> Documents
IEnumerable<IAddressEditable> Addresses

on the IPersonEditable, but then you miss out on AddNew, Remove, Save, Clear, BeginEdit, IsDirty, IsSavable etc... methods wich are available on the collection objects.

it seems like a lot of work to create an IMyCollection interface or something with all that functionality mapped correctly and I was wondering if anyone has tried this already and might give me some pointers or a workaround or maybe some example code.

I was also thinking of maybe solving this (partly) through extensions methods but in those extension methods you would need access to the actual BusinessObject code directly, which would probably be a problem.

regards,
Jurjen.

**Comment by hurcane** [comment]
Have you tried making the Documents and Addresses properties have the actual business object type? Or if you want to stick to interfaces, use one of the other higher-level interfaces provided with CSLA?

**Reply by JurjendeGroot** [reply]
using the actual BO in the interface is not possible, the BO is in a .net 6 project. can't reference that from .NET standard.

**Comment by michaelcsikos** [comment]
Do you have more than one class which implements `IPersonEditable`?

**Reply by JurjendeGroot** [reply]
No just the BusinessObject (and a Fake for testing purposes)

**Reply by michaelcsikos** [reply]
If there's only one implementation of `IPersonEditable` you might not get much benefit from it. If your interfaces can't reference your business objects, then all of your collection properties will need to be declared as interfaces, too, and you can implement some of the interface properties explicitly. Here's a POCO (non-CSLA) example, for simplicity:

```cs
public interface IDocument
{
}

public interface IDocuments
    : IEnumerable<IDocument>
{
}

public interface IPersonEdit
{
    IDocuments Documents { get; }
}

public class Document
    : IDocument
{
}

public class Documents
    : List<Document>
    , IDocuments
{
    IEnumerator<IDocument> IEnumerable<IDocument>.GetEnumerator()
    {
        for (int index = 0; index < Count; ++index)
            yield return this[index];
    }
}

public class PersonEdit
    : IPersonEdit
{
    public Documents Documents { get; private set; }

    IDocuments IPersonEdit.Documents
    {
        get { return Documents; }
    }
}
```

**Comment by rockfordlhotka** [comment]
There are interfaces for each of the base classes in CSLA, specifically to support unit testing or maybe polymorphism.

You can create your interface type as a subclass of the base interface type to get all the other interfaces necessary to act as a CSLA base type (warning: it is a lot).

```c#
public interface IPersonEdit : Csla.IBusinessBase
{
}
```
